### PR TITLE
Android camera provider

### PR DESCRIPTION
**Pull request by jasonxunxu**
This is a fully working camera provider for Android. It is efficient and can achieve pretty high frame rate by receiving frames through a `SurfaceTexture` (based on [this snippet](https://groups.google.com/forum/#!msg/kivy-users/hiyAJqjhFjo/3-WLF4MJGpwJ)).

In addition, it provides an interface to obtain current frame data as a Numpy array (requiring OpenCV Python binding `cv2`), making it suitable for computer vision apps.

It fixes issue #2831. 

This is a recreated PR after rebasing #3236.

### CODE REVIEW

**Comment by kived** [comment]
I don't think you actually rebased this; GitHub says it cannot be automatically merged. Make sure you have pulled the latest changes from Kivy, not just your fork, and then rebase this branch off kivy/master - not the master of your fork. Then do `git push --force` to update the PR.

For convenience, I generally take the approach of making my fork's master branch mirror kivy/master. In my repo I have an 'upstream' remote, so from my master branch I just do `git fetch upstream` then `git reset --hard upstream/master`.

**Comment by kived** [comment]
Thinking about it, your last branch was to merge into the 1.8.0 branch, which means this is probably based on 1.8.0 and not master, and there are thousands of commits in between. It may actually end up being easier to manually redo these changes to a copy of master, rather than trying to rebase.

**Comment by jasonxunxu** [comment]
I actually rebased onto tag 1.8.0 which is on the master branch. I started the development before 1.9.0 was released so just wanted to start with a stable version then. The branch was rebased to tag 1.8.0 instead of current master tip in hope of 1.8.x based forks/branches being able to benefit from it as well.

I tested merging with current master, the only conflict is in `kivy/__init__.py` (where `kivy_options` specifies providers), and quite straightforward to resolve.

**Comment by dessant** [comment]
Avoid basing your prs on 1.9.0 in the future, always use master, sometimes even with master we need to rebase or create a new pr, when conflicting stuff is merged before our code. Great pr btw, this feature will make a lot of users happy. ;)

**Comment by jasonxunxu** [comment]
Got it, thanks for the clarification!

**Comment by tshirtman** [comment]
Actually that wasn't hard to rebase, just had an expected conflict in __init__ (kivy_options), but i think i merged it ok. will do a PR from rebase_3250

**Comment by jasonxunxu** [comment]
Since rebasing is so wanted I will rebase this PR once again onto master. Also I am making some improvements per @tito's [suggestion](https://github.com/kivy/kivy/pull/3236#discussion_r27855620) which will be committed shortly.

**Comment by jasonxunxu** [comment]
Rebased to master, also added a new commit for rendering efficiency improvement as @tito suggested. It seems that merging this PR (instead of  #3255) is easier (thank you @tshirtman for helping though).

**Comment by inclement** [comment]
I guess the reason this hasn't been merged is nobody has tested it? @thirtman, @dessant?

I'll try to look at it this weekend if that's the case. Sorry @jasonxunxu for the delay

**Comment by jasonxunxu** [comment]
Wow, thank you @inclement for lending a hand!

File: kivy/core/camera/camera_android.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+                return None
+            buf = self._buffer.tostring()
+            return buf
+
+    def decode_frame(self, buf):
+        """
+        Decode image data from grabbed frame
+        """
+        import numpy as np
+        from cv2 import cvtColor
```
Comment (akshayaurora) [inline kivy/core/camera/camera_android.py]: This makes numpy and cv2 as a essential requirement on android. Either we should look at replacing these or making bumpy and cv2 as a base requirement for `kivy` recipe in py4a. I assume these are used for performance reasons.

Would it be possible to replace these with another implementation that is equally fast but does not require loading these libs?

**Comment by jasonxunxu** [comment]
It should be clarified that the normal operation of this provider (i.e. showing live stream from the camera) does **not** require OpenCV or Numpy. Those dependencies are only needed when you use the `read_frame` interface to fetch frame data (as a Numpy array) for some processing, in which case you most likely already have cv2 and numpy included in the project.

It would be nice if someone can contribute a fast and dependency-free method to do the needed format conversion, but again using `CameraAndroid` just as a regular camera provider does not make numpy and cv2 essential requirements on Android.

**Comment by tshirtman** [comment]
Thanks for the clarification, i do think this dependency should be documented in the method's docstring.

**Comment by jasonxunxu** [comment]
@tshirtman The dependencies are now documented. A few enhancements were added as well.

**Comment by akshayaurora** [comment]
@jasonxunxu nice work! Really awesome contribution, thanks a ton.

**Pull request merged**